scripts/pi-hole/php/teleporter.php

```diff
@@ -29,68 +31,208 @@ function archive_add_file($path,$name,$subdir="")
+	if(is_null($contents))
+		return 0;
+
+	// Flush table if requested, only flush each table once
+	if($flush && !in_array($table, $flushed_tables))
 	{
-		while(false !== ($entry = readdir($dir)))
+		$db->exec("DELETE FROM ".$table);
+		array_push($flushed_tables, $table);
+	}
+
+	// Prepare field name for domain/address depending on the table we restore to
+	if($table === "adlist")
+	{
+		$sql  = "INSERT OR IGNORE INTO adlist";
+		$sql  .= " (id,address,enabled,date_added,comment)";
+		$sql  .= " VALUES (:id,:address,:enabled,:date_added,:comment);";
+	}
+	elseif($table === "domain_audit")
+	{
+		$sql  = "INSERT OR IGNORE INTO domain_audit";
+		$sql  .= " (id,domain,date_added)";
+		$sql  .= " VALUES (:id,:domain,:date_added);";
+	}
+	else
+	{
+		$sql  = "INSERT OR IGNORE INTO ".$table;
+		$sql  .= " (id,domain,enabled,date_added,comment)";
+		$sql  .= " VALUES (:id,:domain,:enabled,:date_added,:comment);";
+	}
+
+	// Prepare SQLite statememt
+	$stmt = $db->prepare($sql);
+
+	// Return early if we prepare the SQLite statement
+	if(!$stmt)
+	{
+		echo "Failed to prepare statement for ".$table." table.";
+		echo $sql;
+		return 0;
+	}
+
+	// Loop over rows and inject the lines into the database
+	$num = 0;
+	foreach($contents as $row)
+	{
+		$stmt->bindValue(":id", $row["id"], SQLITE3_INTEGER);
+		$stmt->bindValue(":date_added", $row["date_added"], SQLITE3_INTEGER);
+
+		if($table === "adlist")
+			$stmt->bindValue(":address", $row["address"], SQLITE3_TEXT);
+		else
+			$stmt->bindValue(":domain", $row["domain"], SQLITE3_TEXT);
+
+		if($table !== "domain_audit")
 		{
-			if($entry !== "." && $entry !== "..")
-			{
-				archive_add_file($path,$entry,$subdir);
-			}
+			$stmt->bindValue(":enabled", $row["enabled"], SQLITE3_INTEGER);
+			if(is_null($row["comment"]))
+				$type = SQLITE3_NULL;
+			else
+				$type = SQLITE3_TEXT;
+			$stmt->bindValue(":comment", $row["comment"], $type);
+		}
+
+		if($stmt->execute() && $stmt->reset() && $stmt->clear())
+			$num++;
+		else
+		{
+			$stmt->close();
+			return $num;
 		}
-		closedir($dir);
 	}
-}
 
-function limit_length(&$item, $key)
-{
-	// limit max length for a domain entry to 253 chars
-	// return only a part of the string if it is longer
-	$item = substr($item, 0, 253);
+	// Close database connection and return number or processed rows
+	$stmt->close();
+	return $num;
 }
 
-function process_file($contents,$check=True)
+/**
+ * Create table rows from an uploaded archive file
+ *
+ * @param $file object The file of the file in the archive to import
+ * @param $table string The target table
+ * @param $flush boolean Whether to flush the table before importing the archived data
+ */
+function archive_insert_into_table($file, $table, $flush=false, $wildcardstyle=false)
 {
-	$domains = array_filter(explode("\n",$contents));
+	global $db, $flushed_tables;
 
-	// Walk array and apply a max string length
-	// function to every member of the array of domains
-	array_walk($domains, "limit_length");
```

Comment:
This seems to be removed now. If I recall correctly, it prevents a DOS attack via injection of an expensive regex.

Reply:
Should we re-add it? Is it meaningful to protect ourselves against an expensive regex by a mere length limitation?

Reply:
Here is the issue: #464. If we remove it without fixing it in a different way, the issue would have to be re-opened.

Reply:
This is handled by

https://github.com/pi-hole/AdminLTE/pull/986/files#diff-3e979b46422dde1d81341526e434d794R123

I will add the same code in `add_to_table()`.